HTML API: CSS class name methods should behave according to quirks mode

sirreal opened the pull request:
Trac ticket: Core-61531

## Testing changes

These are for the full HTML API test suites. There are no changes to the html5lib test suite.

```diff
- Tests: 2441, Assertions: 4122, Skipped: 423.
+ Tests: 2449, Assertions: 4135, Skipped: 423.
```

## Description

Update that HTML Processor and Tag Processor to handle CSS classes in a case-sensitive way by default.
This aligns with "no-quirks" or "standards" mode behavior for class name handling.
 
Remove forced lowercasing in `::class_list`.

Add a `$document_mode` argument to `WP_HTML_Processor::create_fragment()` to enable the fragment parser to be created in quirks mode.

When the HTML Processor is in no-quirks mode, `add_class`, `remove_class` and `has_class` operate in a _case-sensitive_ way:

```php
$processor = WP_HTML_Processor::create_full_parser( '<!DOCTYPE html><div class="FOO">' );
$processor->next_tag( 'DIV' );
var_dump( $processor->has_class('foo') );
// bool(false)
var_dump( $processor->has_class('FOO') );
// bool(true)
```

When the HTML Processor is in quirks mode (only available via the fragment parser at the moment), `add_class`, `remove_class` and `has_class` operate in a _case-sensitive_ way:

```php
$processor = WP_HTML_Processor::create_fragment('<div class="FOO">', '<body>', 'UTF-8', 'quirks-mode');
$processor->next_tag( 'DIV' );
var_dump( $processor->has_class('foo') );
// bool(true)
var_dump( $processor->has_class('FOO') );
// bool(true)
```

`add_class` and `remove_class` get similar treatment. Case-insensitive classes matching `remove_class( $class_name )` will be removed. Case insensitive duplicate classes will not be added when calling `add_class( $class_name )`.

The case-sensitivity is managed by adding a protected `comparable_class_name` method to the WP_HTML_Tag_Processor. This method is called internally in the CSS class related methods to allow subclasses to customize comparison behavior. This requires minimal changes to the implementation of class name handling in the Tag Processor.

In the Tag Processor and the HTML Processor in "no-quirks" mode, `comparable_class_name` method returns the class name as-is, so case-sensitive comparison is performed. The HTML Processor in quirks mode will return the class name in ASCII lowercase so that case-insensitive comparison is performed.

`class_list` _does_ produce case-insensitive duplicates in all cases. This deduplication would be easy to perform in quirks mode, however it's unclear what form of casing should be yielded. In the case of `<div class="aaa AAA AaA">`, which of the (equivalent in quirks-mode) `aaa` class names should be yielded?

---

The Tag Processor and HTML Processor classes have several methods for dealing with CSS class names: `class_list`, `add_class`, `remove_class` and `has_class`.

These methods are intended to provide a CSS class selector-like interface with the class attribute.

Class name matching ([CSS class selectors `.className {}`](https://www.w3.org/TR/selectors-4/#class-html) or [`getElementsByClassName( "className" )`](https://dom.spec.whatwg.org/#concept-getelementsbyclassname)) is case sensitive in no-quirks mode and case-insensitive in quirks-mode.

> The list of elements with class names classNames for a [node](https://dom.spec.whatwg.org/#concept-node) root is the [HTMLCollection](https://dom.spec.whatwg.org/#htmlcollection) returned by the following algorithm:
> 
> Let classes be the result of running the [ordered set parser](https://dom.spec.whatwg.org/#concept-ordered-set-parser) on classNames.
> If classes is the empty set, return an empty [HTMLCollection](https://dom.spec.whatwg.org/#htmlcollection).
> Return an [HTMLCollection](https://dom.spec.whatwg.org/#htmlcollection) rooted at root, whose filter matches [descendant](https://dom.spec.whatwg.org/#concept-tree-descendant) [elements](https://dom.spec.whatwg.org/#concept-element) that have all their [classes](https://dom.spec.whatwg.org/#concept-class) in classes.
> 
> The comparisons for the [classes](https://dom.spec.whatwg.org/#concept-class) must be done in an [ASCII case-insensitive](https://infra.spec.whatwg.org/#ascii-case-insensitive) manner if root’s [node document](https://dom.spec.whatwg.org/#concept-node-document)’s [mode](https://dom.spec.whatwg.org/#concept-document-mode) is "quirks"; otherwise in an [identical to](https://infra.spec.whatwg.org/#string-is) manner.

> When matching against a document which is in [quirks mode](https://dom.spec.whatwg.org/#concept-document-quirks), class names must be matched [ASCII case-insensitively](https://infra.spec.whatwg.org/#ascii-case-insensitive); class selectors are otherwise case-sensitive, only matching class names they are [identical to](https://infra.spec.whatwg.org/#string-is).

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

dmsnell committed:
Add test setup for running Core tests.

sirreal committed:
Add quirks mode parameter to create_fragment

Allow quirks mode to be set before document processing begins.

sirreal committed:
Add standards mode tests for class name case sensitivity

sirreal committed:
Implement has_class

sirreal committed:
Remove lower-casing behavior of class_list

This is necessary for has_class to work properly.
This could be put into a protected method or the class sensitivity could
be a parameter if desired.

sirreal committed:
Add test with null bytes in class attribute

sirreal committed:
Update to use document_mode

Subsequent changes introduced document_mode instead of compat_mode

sirreal committed:
Replace null bytes in class_list class names

sirreal committed:
Fix tests

sirreal committed:
Handle all the comparison stuff with a protected comparable_class_name function

sirreal committed:
Lint

sirreal committed:
Improve phpdoc explanations

sirreal committed:
Remove comment about styling (HTML structure is affected)

sirreal commented on src/wp-includes/html-api/class-wp-html-processor.php:
```diff
@@ -4645,7 +4675,7 @@ public function remove_class( $class_name ): bool {
 	 */
 	public function has_class( $wanted_class ): ?bool {
-		return $this->is_virtual() ? null : parent::has_class( $wanted_class );
+		return $this->is_virtual() ? false : parent::has_class( $wanted_class );
```
This appears to be a small bug and isn't necessary in this PR. I'd be happy to make another PR if desired. `is_virtual` would suggest we're stopped on a tag, but the tag can't have any attributes. This should return `false`.

dmsnell replied:
hm. I would consider this more of a bug in the type signature. with active format reconstruction it _is_ possible for virtual nodes to have a class, but the `null` was meant to convey exactly what you inferred: "This tag _can have no classes_" rather than "this has _has_ no classes."

dmsnell replied:
Bug introduced in #6753

sirreal replied:
That seems different from what the tag processor does, where "not on a tag" returns `null`, "on a tag" always returns `false` with no special handling for tags with no classes.

https://github.com/WordPress/wordpress-develop/blob/830d66c55cdb5afcaa8c0137d0c0f991258565ee/src/wp-includes/html-api/class-wp-html-tag-processor.php#L1181-L1203

sirreal replied:
> Bug introduced in #6753

I think this was a confusion, there is no bug with types and #6753 did not introduce a bug here that I'm aware of. This is simply a question of what `null` and `false` mean here as return values.

dmsnell replied:
thanks for following up on these comments. `null` has been used to mean "cannot potentially contain that attribute or any others." maybe what we need is a comment update since we created virtual nodes.

at some point, if we allow reading these this might change, as nodes created during active format reconstruction contain the attributes of their original tags. for now, though, I prefer having a distinction between "this tag does not have this class" and "it's not possible to answer if this tag has this class"

sirreal commented on src/wp-includes/html-api/class-wp-html-processor-state.php:
```diff
@@ -393,9 +393,6 @@ class WP_HTML_Processor_State {
 	 *  - When not in `QUIRKS_MODE`, a TABLE start tag implicitly closes an open P tag
 	 *    if one is in scope and open, otherwise the TABLE becomes a child of the P.
 	 *
-	 * `QUIRKS_MODE` impacts many styling-related aspects of an HTML document, but
-	 * none of the other changes modifies how the HTML is parsed or selected.
```
The lines immediately above this about P > TABLE handling in quirks/no-quirks modes seem contradictory. That's directly related to tree-construction.

dmsnell replied:
I've updated the wording. It indeed was self-contradictory

sirreal commented on src/wp-includes/html-api/class-wp-html-processor.php:
```diff
@@ -287,12 +287,14 @@ class WP_HTML_Processor extends WP_HTML_Tag_Processor {
 	 * @return static|null The created processor if successful, otherwise null.
 	 */
-	public static function create_fragment( $html, $context = '<body>', $encoding = 'UTF-8' ) {
+	public static function create_fragment( $html, $context = '<body>', $encoding = 'UTF-8', $document_mode = WP_HTML_Processor_State::NO_QUIRKS_MODE ) {
```
A fragment parser would typically inherit the context element's document's compatibility mode.

I suspect we'll need to change how the context element is passed, but I don't think it will include information about it's document mode and this argument will remain helpful.

See #7141.

dmsnell replied:
do you think this is worth supporting right now? do we have enough warrant to include it? in what cases would we want to create a document fragment in quirks mode, and in those cases, how would we know?

I can definitely see value in having this, but also I wonder if this inclusion will help people understand better what they need to be doing or add confusion.

what do you think the consequences would be of simply not supporting a quirks-mode fragment parser? or at least of not having it in the primary function signature for creating the class?

sirreal replied:
This was helpful and important for testing while working on these changes.

> In what cases would we want to create a document fragment in quirks mode, and in those cases, how would we know?

The fragment should use the same mode as the document for the context element. This should become clear when we work on `set_{inner,outer}_html` where we'll be creating fragment parsers that use the parent parser's document mode.

Another option would be to adjust the full parser to handle doctype declarations and set the document_mode correctly according to the full HTML document. Then we could certainly omit these changes.

I'm also reluctant to eagerly add more to this method signature.

sirreal replied:
> what do you think the consequences would be of simply not supporting a quirks-mode fragment parser? 

If the full parser supports quirks mode and we have `set_{inner,outer}_html` methods, I think the fragment parser must support quirks mode.

> or at least of not having it in the primary function signature for creating the class?

The mode should be based on the context element's document's mode. There's no way to provide that information to the fragment parser right now. The context element is currently passed to the fragment parser as an HTML-like string (only `<body>` is allowed right now) which [seems insufficient to pass all the information](https://github.com/WordPress/wordpress-develop/pull/7141#issuecomment-2283867964) the fragment parser requires.

There are other ways to handle this, for example an instance method could create a fragment from a node and set quirks mode appropriately as well as handle things like reading attributes, namespace, etc. This is all best discussed in #7141.

I am going to explore a change to handle quirks mode in the full parser based on the doctype declaration, that would be sufficient and we could remove this change.

dmsnell replied:
I could see some value in having a method like `->set_quirks_mode()`

I still suspect that it'll be basically universal that we don't work in the full parser mode within WordPress. almost no HTML-processing code has access to the full document, and so that's what I meant when assuming _no-quirks mode_. We just don't know even when doing `inner_html` operations if the parent document is in quirks mode or not. "We can only assume UTF-8, no-quirks, `<body>` context."

sirreal replied:
There may be value in allowing the compat mode to be changed on a processor. I'd like to leave those changes for consideration in their own PR. There was no way to change the compat mode before and I don't think we need to address that in this PR.

sirreal commented on src/wp-includes/html-api/class-wp-html-tag-processor.php:
```diff
@@ -1155,12 +1157,7 @@ public function class_list() {
-			 * @see https://www.w3.org/TR/CSS2/syndata.html#x1
-			 */
-			$name = strtolower( substr( $class, $at, $length ) );
+			$name = str_replace( "\x00", "\u{FFFD}", substr( $class, $at, $length ) );
```
**This needs to be extracted to its own PR for consideration.**

The browser will replace null bytes with `�` in attribute values. It's impossible to have a null byte in an attribute value that comes from HTML. It's possible to set a null byte in an attribute value via the DOM.

We may consider that `class_list` is a "processed" version of the `class` attribute, and this null-byte replacement is appropriate here. At this time, `get_attribute` is closer to "raw" access to the attribute value as authored.

Related to https://github.com/WordPress/wordpress-develop/pull/7184.

sirreal replied:
Moved this to https://github.com/WordPress/wordpress-develop/pull/7187.

dmsnell replied:
I don't think this needs to be in a separate PR. we can entertain it wherever. A separate PR is fine though. The focus is on handling CSS class names properly, so in that regard this is purely about matching CSS class selectors to values in HTML `class` attributes.

sirreal committed:
Revert "Replace null bytes in class_list class names"

This reverts commit 4f1836818ca863f0611409fce5658acbe3fffe12.

sirreal committed:
Remove null-byte has_class test

dmsnell commented on src/wp-includes/html-api/class-wp-html-tag-processor.php:
```diff
@@ -1191,10 +1187,10 @@ public function has_class( $wanted_class ): ?bool {
 
 		foreach ( $this->class_list() as $class_name ) {
-			if ( $class_name === $wanted_class ) {
+			if ( $this->comparable_class_name( $class_name ) === $wanted_class ) {
```
in a case like this it seems like it would be better to extract the same logical condition check outside the `foreach` loop rather than repeat it on every iteration. we know that the quirks mode won't change during the loop, so we could create two loops - one for each mode.

dmsnell replied:
casing here is the only consequence and "is quirks mode" is the only check, right? so could we not inline this with `substr_compare()` and avoid the function call?

```php
$wanted_length    = strlen( $wanted_class );
$case_insensitive = self::QUIRKS_MODE === $this->compatibility_mode;

foreach ( $this->class_list() as $class_name ) {
	if (
		strlen( $class_name ) === $wanted_length &&
		0 === substr_compare( $class_name, $wanted_class, 0, $wanted_length, $case_insensitive )
	) {
		return true;
	}
```

sirreal replied:
We can do that, but if take that approach it may be best to inline the `comparable_class_name` logic everywhere so that it's not assumed that function is not applied everywhere it's needed.

I'll push those changes and we can decide if that's preferable.

sirreal replied:
I've pushed 15d479e and 8dbf6ab to remove the `comparable_class_name` method.

dmsnell commented on src/wp-includes/html-api/class-wp-html-tag-processor.php:
```diff
@@ -3853,12 +3849,27 @@ public function remove_class( $class_name ): bool {
+	 * @param string $class_name The class name to transform.
+	 * @return string The transformed class name.
+	 */
+	protected function comparable_class_name( string $class_name ): string {
```
why not create this as an internal method and add the quirks-mode flag to the Tag Processor?

sirreal replied:
That would simplify some things, happy to make that change if you prefer.

I didn't add it because it's an HTML concept and seems more appropriate for the HTML processor. Those lines have already started to blur.

dmsnell replied:
in this case I see this more as a CSS concept and I don't see a reason to make the Tag Processor less adequate. CSS class name matching was always a concept we incorporated because of the obvious need to use it. the CSS class name matching has always had custom code.

I would rather in this case embed the quirks mode check in the Tag Processor, or at least assume _no-quirks_ mode by default in the absence of that. This _does_ imply that the quirks mode flag should live in the Tag Processor, but that seems reasonable since normal HTML parsing won't examine it.

sirreal commented:
I plan to revisit this when https://github.com/WordPress/wordpress-develop/pull/7195 is complete. That will allow us to test with quirks mode using the full parser and not need to change the fragment factory method signature.

sirreal committed:
Merge branch 'trunk' into html-api/css-class-name-method-audit

sirreal committed:
Revert ::create_fragment changes

Full documents can be created in quirks mode now. There's no need to
introduce quirks mode to the fragment parser or change its signature in
order to tests the quirks mode changes.

sirreal committed:
Adjust tests to use full parser in quirks/no-quirks

sirreal committed:
Move quirks mode to tag processor

Quirks mode changes behavior CSS class functions, namely whether they
are ASCII case-insensitive class name matches or byte-for-byte
comparisons.

It makes sense to move quirks mode into the tag processor so that it can
deal with this correctly.

sirreal committed:
Make comparable_class_name internal

sirreal commented:
I've moved quirks mode into the Tag Processor, it impacts CSS class name handling and it makes sense to have that there.

Now that [#7195](https://github.com/WordPress/wordpress-develop/pull/7195) / [[58925]](https://core.trac.wordpress.org/changeset/58925) has landed, changes in quirks mode handling are tested by using the full HTML parser in quirks mode and no-quirks mode.

Quirks mode is not exposed publicly at this time (no change from current `trunk`). In a previous commit, an argument was added to `::create_fragment` to adjust the compat mode. This is no longer necessary and has been reverted.

dmsnell commented on src/wp-includes/html-api/class-wp-html-tag-processor.php:
```diff
@@ -511,6 +511,18 @@ class WP_HTML_Tag_Processor {
+	 *
+	 * @var string
+	 */
+	protected $compat_mode = self::NO_QUIRKS_MODE;
```
Can we not bring over the previous comment here, and update it with the correction to the contradictory statement you mentioned?

sirreal replied:
Is bb6d772 sufficient or do you think there should be more information on the property?

dmsnell replied:
having it on the constants is a good requirement. I can still see value in having at least a summary here, though I know we're not consistent everywhere in this. what I'm thinking about is someone hovering over the `$compat_mode` property and seeing documentation appear inline. for them, having a basic summary could help quickly inform them of whether or not they need to care about this. some IDEs do make it easy to follow the `@see` links, but not all do, and source code readers don't.

let me stew on it. maybe I can work at it some as well

dmsnell replied:
I've brought the comment over and updated the wording to make it clearer.

wanted to copy it here so that when people hover over `compat_mode` they get an instant clue on what it is and how it affects things

dmsnell commented on src/wp-includes/html-api/class-wp-html-tag-processor.php:
```diff
@@ -1126,6 +1138,7 @@ public function paused_at_incomplete_token(): bool {
 	 *     // Outputs: "free <egg> lang-en "
 	 *
 	 * @since 6.4.0
+	 * @since 6.7.0 Class names are no longer force lower-cased.
```
How about something affirmative like…

```php
/**
 * @since 6.7.0 Class name casing corresponds to document compatibility mode.
 */
```

sirreal replied:
I'm open to changing this, but if it's included I think it should be clear what's changed.

Maybe this is a bugfix and does not need a since annotation at all.

sirreal replied:
I've removed the since tag.

dmsnell commented on src/wp-includes/html-api/class-wp-html-tag-processor.php:
```diff
@@ -1155,12 +1168,7 @@ public function class_list() {
-			 * @see https://www.w3.org/TR/CSS2/syndata.html#x1
-			 */
-			$name = strtolower( substr( $class, $at, $length ) );
+			$name = substr( $class, $at, $length );
```
why not go ahead and base this off of the document compatibility mode?

that is, lower-case it if we're in quirks mode…I think it's fine to return the lower-case version in this case because of it being comparable.

dmsnell replied:
okay I see. well, for adding it to the `$seen` array I can see my comment being valid. lower-case the seen list so we can filter duplicates, but reports the casing of the first match as the casing of the attribute. this preserves author intent as far as we can make it while being reasonable.

sirreal replied:
I'll state the problem for clarity. In quirks mode we can deduplicate class name case-insensitively, but then what form should be yielded? Lowercase (this was the behavior before)? Uppercase? First/last/nth occurrence in HTML? For example, all the class names in this HTML are identical CSS class names, so if _one_ is yielded, which?

```html
<!DOCTYPE this-will-be-quirks-mode>
<div class="SaMe sAmE SAME same">
```

---

If we perform deduplication and yield in quirks mode, it doesn't seem like there's any way to get it right. It really feels to me like we should just yield all the unique class names and the fact that repeated class names are yielded is something that calling code can deal with if they desire.

This will only affect quirks mode, which is expected to be the less common mode. The Tag Processor and HTML (Fragment) Proccessor will operate in no-quirks mode by default.

dmsnell replied:
> It really feels to me like we should just yield all the unique class names and the fact that repeated class names are yielded is something that calling code can deal with if they desire.

this is the only situation I truly feel strongly about, and it's that we shouldn't do it, because of problems like the fundamental question: "how many class names are here?" we would be making it imperative for calling code to perform CSS-level semantic parsing when interacting with our CSS functions.

it's possible we're overthinking it. I see value in lower-casing them for normalization purposes (people will likely be checking with lower-case variations in their calling code anyway, even though `has_class()` is here for checking for the presence of a class name).

the reason for reporting the case as-is, for the first class name listed, is that in a way it preserves the author intent. in these situations we have to realize that the markup is invalid in some way. how much do we fix invalid code? well, we can't, but it's our call how to present these situations.

in `set_attribute()` we preserve the casing given for the attribute. maybe this should only be a principle for _setting_ functions and let _getting_ functions read the normalized form (which is how `get_attribute()`) works. in fact, `get_attribute_names_with_prefix()` lower-cases attribute names, so maybe the best option is matching that behavior, where `class_list()` in quirks-mode reports lower-cased variants, but `has_class()` always works, and `add_class()` preserves the provided casing.

sirreal replied:
Ok, I've updated `class_list` to yield lowercased and appropriately de-duplicated class names in quirks mode.

dmsnell commented on src/wp-includes/html-api/class-wp-html-tag-processor.php:
```diff
@@ -4248,4 +4280,29 @@ public function get_doctype_info(): ?WP_HTML_Doctype_Info {
+	 * No-quirks mode document compatability mode.
+	 *
+	 * > In no-quirks mode, the behavior is (hopefully) the desired behavior
+	 * > described by the modern HTML and CSS specifications.
```
whence comes this quote? and what does "hopefully" communicate to someone trying to figure out the purpose of this mode?

dmsnell replied:
even though the Tag Processor doesn't parse HTML differently because of these, it would be nice to still discuss the impact of TABLE elements closing an open P. these constants will still provide hover-documentation inside the HTML Processor code, and could inform someone, who is poking around and reading the docs, that this has such an impact.

sirreal replied:
> whence comes this quote? and what does "hopefully" communicate to someone trying to figure out the purpose of this mode?

This PR moves this, it was introduced in Processor State in https://github.com/WordPress/wordpress-develop/pull/6972 / [[58779]](https://core.trac.wordpress.org/changeset/58779). I'm not sure where it came from originally.

> even though the Tag Processor doesn't parse HTML differently because of these, it would be nice to still discuss the impact of TABLE elements closing an open P. these constants will still provide hover-documentation inside the HTML Processor code, and could inform someone, who is poking around and reading the docs, that this has such an impact.

I struggled with how to document different behavior without being confusing or too length. I'll add some notes again and see what you think.

sirreal replied:
See bb6d772ff188fc5b765a9dbf2808a89bfd2503c6 for notes about quirks-mode.

dmsnell replied:
Found it; this is why I need review and scrutiny on my own contributions. 🙃

https://developer.mozilla.org/en-US/docs/Web/HTML/Quirks_Mode_and_Standards_Mode

sirreal committed:
Add information about how quirks mode changes behavior

sirreal committed:
Remove comparable_class_name function call from has_class loop

sirreal committed:
Remove comparable_class_name method entirely

sirreal committed:
Add another test for quirks-mode add_class

sirreal committed:
Remove since tag from class_list

sirreal committed:
Lowerclass yielded class_list class names in quirks mode

sirreal committed:
Fix modifying class case when removing another class

sirreal commented:
I think the concerns have been addressed and this is ready for another review.

sirreal committed:
Merge branch 'trunk' into html-api/css-class-name-method-audit

sirreal committed:
Fix equals sign alignment lint

dmsnell committed:
Merge branch 'trunk' into html-api/css-class-name-method-audit

dmsnell committed:
Reintroduce an explanatory comment on `compat_mode` property

dmsnell committed:
Revert: Have `has_class()` return `null` for unsupported tokens.

dmsnell commented on src/wp-includes/html-api/class-wp-html-processor.php:
```diff
@@ -4938,6 +4938,10 @@ public function remove_class( $class_name ): bool {
+	 *
 	 * @param string $wanted_class Look for this CSS class name, ASCII case-insensitive.
 	 * @return bool|null Whether the matched tag contains the given class name, or null if not matched.
 	 */
```
@sirreal I've reverted this change so we can consider it separately. I know it will be important during active format reconstruction, but I think that `null` is a kind of partially-implemented escape hatch that communicates that this isn't supported rather than indicating that a class definitively doesn't exist on the tag.

dmsnell committed:
Truncate explanatory comment on quirks mode constants and cite MDN.

dmsnell commented on tests/phpunit/tests/html-api/wpHtmlProcessor.php:
```diff
@@ -519,4 +519,132 @@ public function test_foreign_content_script_self_closing() {
+		$this->assertSame( '<span class="UPPER">', $processor->get_updated_html() );
+
+		$processor->add_class( 'ANOTHER-UPPER' );
+		$this->assertSame( '<span class="UPPER another-upper">', $processor->get_updated_html() );
```
I was hoping that here we could respect the given casing.

dmsnell committed:
Modify tests

dmsnell committed:
Preserve given casing of added and removed class names.

dmsnell commented:
@sirreal I've updated some comments and changed the behavior of `add_class()` and `remove_class()` so that they preserve the casing of the first-provided class name for all lexical variations. I'll entertain ideas on this, but it's how we handle attributes, and I think it's the most respectful we can be to developers who are setting values when casing doesn't matter.

it's really quite a surprise, and I hope quirks mode is almost never used, given the conflict between CSS selectors matching the `class` attribute value, and those given as class name selectors.

dmsnell committed:
Merge branch 'trunk' into html-api/css-class-name-method-audit

dmsnell reviewed (approved):
@sirreal I'm going to merge this because I anticipate you would prefer it (once the tests all pass), but if you disagree with the changes I made we can revisit.

dmsnell committed:
Merge branch 'test-everything' into html-api/css-class-name-method-audit

dmsnell committed:
Fix issue with classname updates, and change test to assert it.

dmsnell committed:
Remove test helpers accidentally added (can't force-push)

dmsnell commented:
@sirreal I've also moved the Trac ticket reference to the top of the description to make it easier to find, and used the shorthand notation instead of the full link.

dmsnell commented:
Merged in [[58985]](https://core.trac.wordpress.org/changeset/58985)
https://github.com/wordpress/wordpress-develop/commit/fb40fe915efab59d54d5bf812e2a4574ae525510
